### PR TITLE
Collect real estates rental prices

### PR DESCRIPTION
**Pull request by Lrcezimbra**
Collect real estates rental prices from [ZAP Imóveis](http://www.zapimoveis.com.br/busca-mapa). It solves the problem of granularity cited on #27 and #20. 

Now it's collecting data from Distrito Federal, but can easily get from other locations. 
The collected data is: 
`ID, coordinates, rent_rate, sale_rate, rental_price, sale_price, total_area, useful_area, neighborhood, dorms, subtype, suite, vacancies`

I removed some columns that I think it's not necessary for the project.

### CODE REVIEW

File: src/fetch_real_estate_prices.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+import json
+import numpy as np
+import requests
+import os
+import pandas as pd
+from geopy.geocoders import GoogleV3
```
Comment (cuducos) [inline src/fetch_real_estate_prices.py]: I can't find this module `geopy`. Should we add it to `conda_requirements.txt`?

File: src/fetch_real_estate_prices.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
+import json
```
Comment (cuducos) [inline src/fetch_real_estate_prices.py]: Disclaimer: 💅

Python style guide propose [a convention to organize the `import` section](https://www.python.org/dev/peps/pep-0008/#imports). In a nutshell first we have native modules, then modules installed via `pip`, and then your application's modules. And on top of that, `import` comes before `from … import …`.

That said, I would have written:

```python
import json
import os

import numpy as np
import pandas as pd
import requests
from geopy.geocoders import GoogleV3
```

But… well… again: 💅

File: src/fetch_real_estate_prices.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import numpy as np
+import requests
+import os
+import pandas as pd
+from geopy.geocoders import GoogleV3
+
+HTTP_HEADERS = {"User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/32.0.1700.102 Safari/537.36"}
+OUTPUT_FILEPATH = os.path.join('data', 'real_estates_prices.csv')
```
Comment (cuducos) [inline src/fetch_real_estate_prices.py]: Your PR's made me realize we have been using a kind of standard for the `/data` that was undocumented. We are prefixing all files with date (`YYYY-MM-DD-my-dataset.xz`) and we are using `.xz` ([lzma](https://docs.python.org/3/library/lzma.html)) compression.

Would you mind updating the naming and format of the output file?

File: src/fetch_real_estate_prices.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+LOCATION = 'Distrito Federal, Brasil'
```
Comment (cuducos) [inline src/fetch_real_estate_prices.py]: I'm wondering: how would this work for fetching data from all states?

File: src/fetch_real_estate_prices.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+def slice_quadrants(northeast, southwest, size=0.0625):
```
Comment (cuducos) [inline src/fetch_real_estate_prices.py]: If we're dealing with tons of data, using generators might be a wise choice:

```python
def slice_quadrants(northeast, southwest, size=0.0625):
    for lat in np.arange(southwest['lat'], northeast['lat'], size):
        for long in np.arange(southwest['lng'], northeast['lng'], size):
            yield ((str(lat), str(long)), (str(lat + size), str(long + size)))
```

**Comment by cuducos** [comment]
One more thing: you could document the purpose and usage of the script in `CONTRIBUTING.md` ; )

**Comment by Lrcezimbra** [comment]
@cuducos thanks for the code review and tips. I'll work on this improvements.

File: src/fetch_real_estate_prices.py (hunk 1 above)
Comment (Lrcezimbra) [inline src/fetch_real_estate_prices.py]: Already is in `conda_requirements.txt` line 2 `geopy==1.11.0`

File: src/fetch_real_estate_prices.py (hunk 3 above)
Comment (Lrcezimbra) [inline src/fetch_real_estate_prices.py]: Fixed on 4e085d3

File: src/fetch_real_estate_prices.py (hunk 4 above)
Comment (Lrcezimbra) [inline src/fetch_real_estate_prices.py]: Change to a list in b0bcc02 to support multiple locations.

File: src/fetch_real_estate_prices.py (hunk 5 above)
Comment (Lrcezimbra) [inline src/fetch_real_estate_prices.py]: Good idea, implemented in b501c93.

File: src/fetch_real_estate_prices.py (hunk 2 above)
Comment (Lrcezimbra) [inline src/fetch_real_estate_prices.py]: I was in doubt about how to organize the imports, but didn't search about. Thanks for the tip.

Fixed in 0b48e05. But noting now, requests is a native module, correct? I didn't pay attention to this.

**Comment by Lrcezimbra** [comment]
I pushed some commits. Still missing:

- [ ] docstrings
- [ ] CONTRIBUTING.md
- [ ] lines too long

And I'm adding threads to run the script faster.
- [ ] Threads

I'll work on this asap.

File: src/fetch_real_estate_prices.py (hunk 2 above)
Comment (cuducos) [inline src/fetch_real_estate_prices.py]: [`resquests`](http://docs.python-requests.org/en/master/) isn't native but I guess it's installed by default in Anaconda Python. The native module is [`urllib.request`](https://docs.python.org/3.5/library/urllib.request.html#module-urllib.request) — easy to mess up, isn't it? ; )

**Comment by cuducos** [comment]
That TODO list, and the improvements, are awesome! Congrats and thank you!

Let me know if you need some help and I'd be happy to pair ; )

**Comment by Lrcezimbra** [comment]
I've been kind of busy in the lasts weeks. Next week I'll finish the pull request.

**Comment by Lrcezimbra** [comment]
Add docstring in b92609f. My English isn't very good, so if my docstrings aren't good, please let me know that I try to improve.

I think this latest commits solves all the fixes and improvements cited by @cuducos

**Comment by cuducos** [comment]
For some reason I completely forgot to get back to this PR (probably because real state price was out of our first hypothesis)!

Sorry @Lrcezimbra!

Before merging I'd like to see two minor changes:

* Can you update `CONTRIBUTING.md` too avoid conflicts? If you want, I already [fixed conflicts and typos/wording](https://gist.github.com/a033205f36114b550591b61670ac076f)
* Can you extract `location` with `sys.argv` or `argparse`? So instead of editing the `.py` file, one just add the location in the terminal (eg `$ python src/fetch_real_estate_prices.py "Araguari, Minas Gerais"`)

**Comment by Lrcezimbra** [comment]
I'll work on it.

**Comment by cuducos** [comment]
Closed by inactivity – surely anyone is free to jump in and take this great contribution further ; )

**Pull request closed without merging**